autogpt/processing/text.py

```diff
@@ -1,170 +1,234 @@
+            "Do not directly answer the question itself"
+        )
 
-        messages = [create_message(chunk, question)]
-        tokens_for_chunk = count_message_tokens(messages, model)
-        logger.info(
-            f"Summarizing chunk {i + 1} / {len(chunks)} of length {len(chunk)} characters, or {tokens_for_chunk} tokens"
+    summarization_prompt = ChatPrompt.for_model(model)
+
+    token_length = count_string_tokens(text, model)
+    logger.info(f"Text length: {token_length} tokens")
+
+    # reserve 50 tokens for summary prompt, 500 for the response
+    max_chunk_length = _max_chunk_length(model) - 550
+    logger.info(f"Max chunk length: {max_chunk_length} tokens")
+
+    if not must_chunk_content(text, model, max_chunk_length):
+        # summarization_prompt.add("user", text)
+        summarization_prompt.add(
+            "user",
+            "Write a concise summary of the following text"
+            f"{f'; {instruction}' if instruction is not None else ''}:"
+            "\n\n\n"
+            f'LITERAL TEXT: """{text}"""'
+            "\n\n\n"
+            "CONCISE SUMMARY: The text is best summarized as"
+            # "Only respond with a concise summary or description of the user message."
         )
 
+        logger.debug(f"Summarizing with {model}:\n{summarization_prompt}\n")
         summary = create_chat_completion(
-            model=model,
-            messages=messages,
-        )
-        summaries.append(summary)
-        logger.info(
-            f"Added chunk {i + 1} summary to memory, of length {len(summary)} characters"
+            summarization_prompt, model, temperature=0, max_tokens=500
         )
 
-        memory_to_add = f"Source: {url}\n" f"Content summary part#{i + 1}: {summary}"
+        logger.debug(f"\n{'-'*16} SUMMARY {'-'*17}\n{summary}\n{'-'*42}\n")
+        return summary.strip(), None
 
-        memory.add(memory_to_add)
+    summaries: list[str] = []
+    chunks = list(split_text(text, for_model=model, max_chunk_length=max_chunk_length))
+
+    for i, (chunk, chunk_length) in enumerate(chunks):
+        logger.info(
+            f"Summarizing chunk {i + 1} / {len(chunks)} of length {chunk_length} tokens"
+        )
+        summary, _ = summarize_text(chunk, instruction)
+        summaries.append(summary)
 
-    logger.info(f"Summarized {len(chunks)} chunks.")
+    logger.info(f"Summarized {len(chunks)} chunks")
 
-    combined_summary = "\n".join(summaries)
-    messages = [create_message(combined_summary, question)]
+    summary, _ = summarize_text("\n\n".join(summaries))
 
-    return create_chat_completion(
-        model=model,
-        messages=messages,
-    )
+    return summary.strip(), [
+        (summaries[i], chunks[i][0]) for i in range(0, len(chunks))
+    ]
 
 
-def scroll_to_percentage(driver: WebDriver, ratio: float) -> None:
-    """Scroll to a percentage of the page
+def split_text(
```

Comment:
This is a wildly complex function with very high cyclomatic complexity that should be split into smaller chunks.

Reply:
True, it is pretty complex. Is it an issue of readability? In that case, wouldn't adding comments where appropriate be enough? Because I only see 4 lines in the function that would be useful as a separate function.

Reply:
Gotta run, but can make suggestions later

Reply:
Okay, I've run out of time again, so I think you go with it.  The real root of the complexity in this function is that you're working with your text in unencoded space and your metric for splitting (token count) in encoded space and so you're having to juggle a lot of state and  make regular conversions between the two.  I think most of the logic in this entire module can get a lot cleaner if you moved into token space at the beginning of this text splitting and only converted back to unencoded space at the very end.  

I think you can merge this as is and either you or I can PR against it later.